# **Project Plan**

Hello! Thank you for reviewing my project plan. I was unclear how much analysis I was supposed to do on this plan submission, so I haven't done much exploration yet. I have detailed all the exploration I plan to do. I look forward to your feedback on getting this plan where it needs to be. Thank you :)

## **Import Packages and Data**

In [19]:
import numpy as np
import pandas as pd

In [12]:
# Read in dataframes
try:
    contract_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/contract.csv', parse_dates=True)
    internet_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/internet.csv', parse_dates=True)
    personal_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/personal.csv', parse_dates=True)
    phone_df = pd.read_csv('/Users/kellyshreeve/Desktop/Data-Sets/Final Project/final_provider/phone.csv', parse_dates=True)
except:
    contract_df = pd.read_csv('/datasets/contract.csv', parse_dates=True)
    internet_df = pd.read_csv('/datasets/internet.csv', parse_dates=True)
    personal_df = pd.read_csv('/datasets/personal.csv', parse_dates=True)
    phone_df = pd.read_csv('/datasets/phone.csv', parse_dates=True)

In [13]:
# Print df info
print('Contract DF')
print()
print(contract_df.info())
print()

contract_df.head()

Contract DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   BeginDate         7043 non-null   object 
 2   EndDate           7043 non-null   object 
 3   Type              7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
dtypes: float64(1), object(7)
memory usage: 440.3+ KB
None



customerID   BeginDate              EndDate            Type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  
0              Yes           Electronic check           29.85        29.85  
1               No               Mailed check           56.95       1889.5  
2              Yes               Mailed check           53.85       108.15  
3               No  Bank transfer (automatic)           42.30      1840.75  
4              Yes           Electronic check           70.70       151.65

In [14]:
print('Internet DF')
print()
print(internet_df.info())
print()

internet_df.head()

Internet DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5517 entries, 0 to 5516
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   customerID        5517 non-null   object
 1   InternetService   5517 non-null   object
 2   OnlineSecurity    5517 non-null   object
 3   OnlineBackup      5517 non-null   object
 4   DeviceProtection  5517 non-null   object
 5   TechSupport       5517 non-null   object
 6   StreamingTV       5517 non-null   object
 7   StreamingMovies   5517 non-null   object
dtypes: object(8)
memory usage: 344.9+ KB
None



customerID InternetService OnlineSecurity OnlineBackup DeviceProtection  \
0  7590-VHVEG             DSL             No          Yes               No   
1  5575-GNVDE             DSL            Yes           No              Yes   
2  3668-QPYBK             DSL            Yes          Yes               No   
3  7795-CFOCW             DSL            Yes           No              Yes   
4  9237-HQITU     Fiber optic             No           No               No   

  TechSupport StreamingTV StreamingMovies  
0          No          No              No  
1          No          No              No  
2          No          No              No  
3         Yes          No              No  
4          No          No              No

In [15]:
print('Personal DF')
print() 
print(personal_df.info())
print()

personal_df.head()

Personal DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     7043 non-null   object
 1   gender         7043 non-null   object
 2   SeniorCitizen  7043 non-null   int64 
 3   Partner        7043 non-null   object
 4   Dependents     7043 non-null   object
dtypes: int64(1), object(4)
memory usage: 275.2+ KB
None



customerID  gender  SeniorCitizen Partner Dependents
0  7590-VHVEG  Female              0     Yes         No
1  5575-GNVDE    Male              0      No         No
2  3668-QPYBK    Male              0      No         No
3  7795-CFOCW    Male              0      No         No
4  9237-HQITU  Female              0      No         No

In [17]:
print('Phone DF')
print()
print(phone_df.info())
print()

phone_df.head()

Phone DF

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6361 entries, 0 to 6360
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customerID     6361 non-null   object
 1   MultipleLines  6361 non-null   object
dtypes: object(2)
memory usage: 99.5+ KB
None



customerID MultipleLines
0  5575-GNVDE            No
1  3668-QPYBK            No
2  9237-HQITU            No
3  9305-CDSKC           Yes
4  1452-KIOVK           Yes

Notes on data frames in the project plan below.

## **Examine Class Balance**

In [24]:
# Create churn feature where 0 = No and 1 = Yes
contract_df['churn'] = np.where(contract_df['EndDate'] == 'No', 0, 1)

contract_df.head()

customerID   BeginDate              EndDate            Type  \
0  7590-VHVEG  2020-01-01                   No  Month-to-month   
1  5575-GNVDE  2017-04-01                   No        One year   
2  3668-QPYBK  2019-10-01  2019-12-01 00:00:00  Month-to-month   
3  7795-CFOCW  2016-05-01                   No        One year   
4  9237-HQITU  2019-09-01  2019-11-01 00:00:00  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

   churn  
0      0  
1      0  
2      1  
3      0  
4      1

In [34]:
class_balance = contract_df['churn'].value_counts().reset_index()
class_balance['percent'] = ((class_balance['count'] / len(contract_df)) * 100).round(2)

print('Customer Churn Class Balance')
print(class_balance)

Customer Churn Class Balance
   churn  count  percent
0      0   5174    73.46
1      1   1869    26.54


* The dataset does have an imbalance, with only 26.54% customers having churned and 73.46% of customers with active subscriptions.
* While imbalanced, this is likely not extreme enough to require class balancing techniques.

## **Project Plan**

This is a binary classification, supervised machine learning problem, where the goal is to predict whether customers are likely to churn. The analysis will move from data preparation, to exploratory analysis, and finish with training and validating classification machine learning models. The goal is to maximize AUC-ROC.

#### **Data Frames**  

Contract DF:
* Dataset contains customers' plan start and end date, type of plan, billing type, payment method, and charges.
* Size: 7043 observations and 8 columns. 
* Missing Values: There are no np.nan values, though further investigation is needed, as there may be missing values recoreded in another way. 
* Data Types: Total charges is read in as object. There may be string entries in the column needing cleaned so the column can be properly recorded as float. Dates have not been properly parsed and should be explicitly defined as datetime.

Internet DF:
* Dataset has the type of internet service and whether a customer has added extra protection or streaming services.
* Size: 5517 observations and 8 columns. 
* Missing Values: There are no np.nan values, though further investigation is needed, as there may be missing values recoreded in another way.
* Data Types: All data types are read in correctly as object.

Personal DF:
* Dataset has information on customers' demographics, including gender, senior status, partnership, and dependents.
* Size: 7043 observations and 5 columns. 
* Missing Values: There are no np.nan values, though further investigation is needed, as there may be missing values recoreded in another way.
* Data Types: All data types are correct.

Phone DF:
* Dataset has infomormation on whether customers have one or two phone lines
* Size: 6361 observations and 2 columns.
* Missing Values: There are no np.nan values, though further investigation is needed, as there may be missing values recoreded in another way.
* Data Types: All data types are correct.

#### **Data Analysis Plan**

**Data Preparation**
* Convert column names to snake case.
* Combine all three datasets with an outer merge on the 'cutomer_id' column. Customers who do not have a service will be filled with np.nan for those columns.
* Fill the np.nan cells that were created from the merge with a new category called 'NA'. This will represent customers who do not have phone or internet service and will allow those customers to be kept in the model.  

**Feature Creation**
* 'churned' where 0 = has not churned and 1 = has churned
* 'contract_length' containing the number of days from beginning to end of contract, for those who have churned, and the number of days from beginning of contract to latest date in the dataset, for those who have not churned.
* 'begin_year' containing the year of contract start.
* 'total_internet_services' summing the total number of extra internet services a customer subscribes to (online_security through streaming_movies).

**Exploratory Analysis**
* Construct frequency tables and bar charts for each categorical feature. Examine variable levels and fix incorrect categories and implicit duplicates.
* Construct histograms for each quantitative feature. Examine shape, center, and spread. Watch for outliers and eliminate if necessary.

**Build Models**  

Target:   
* **churned**: 0 = has not churned, 1 = has churned

Features:   
* **contract_length**: number of days from start to end, for those who churned, or number of days from start to last date in the dataset, for those who haven't
* **begin_year**: year of contract start date
* **type**: type of contract  
* **paperless_billing**: customer has paperless billing or not  
* **payment_method**: type of payment method  
* **monthly_charges**: amount customer is charged per month  
* **total_charges**: total charges customer has to date  
* **internet_service**: type of internet service  
* **online_security**: whether customer has this service, 'NA' for customers who do not have internet  
* **online_backup**: whether customer has this service, 'NA' for customers who do not have internet  
* **device_protection**: whether customer has this service, 'NA' for customers who do not have internet  
* **tech_support**: whether customer has this service, 'NA' for customers who do not have internet  
* **streaming_tv**: whether customer has this service, 'NA' for customers who do not have internet  
* **streaming_movies**: whether customer has this service, 'NA' for customers who do not have internet  
* **total_internet_services**: count of extra internet services to which the customer has subscribed  
* **gender**: male or female  
* **senior_citizen**: whether the customer is a senior citizen  
* **partner**: whether the customer has a partnmer  
* **dependents**: whether the customer has dependents  
* **multiple_lines**: whether a customer has multiple phone lines, 'NA; for customers who do not have phone 

Models:
* Logistic Regression
    - One-Hot Encode categorical features
    - Scale monthly and total charges

* Classification Gradient Boosting
    - CatBoost
    - LightGBM
    - XGBoost

Model evaluation: 
* Split data into training, validation, and test-set. 
    - Logistic regression will be trained on training and validated on validation set
    - Gradient Boosting models will be cross-validated on the combined training and validation sets using GridSearchCV and Pipelines to avoid data leakage
* AUC-ROC, Accuracy, and F1-Score
    - Goal is to maximize AUC-ROC
    - Both accuracy and F1 score will be reported as there is a slight class imbalance

**Questions**
* If a customer is not listed on the internet or phone dataset, can it be assumed they do not have those services?
* Do this data warrant class balancing techniques?